ytt: init at 0.30.0

brhoades opened the pull request:
https://github.com/vmware-tanzu/carvel-ytt

###### Motivation for this change
ytt is not already available in nixpkgs. Tests ran by checkPhase fail due to YAML marshaling issues but failures are unrelated to the result binary.

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- N/A  Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

wilsonehusin reviewed (approved):
LGTM

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -1252,6 +1252,16 @@
     githubId = 3043718;
     name = "Brett Lyons";
   };
+  brodes = {
```
Please put this into a separate commit with the message ``maintainers: add brodes``.

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+{ pkgs, lib, ... }:
+with pkgs; with lib; buildGoModule rec {
```
```suggestion
{  lib }:
buildGoModule rec {
```

Don't use with for the entire scope. Also don't pass in pkgs. Triple dots are only needed if you can add any number of extra packages which is not needed here.

brhoades replied:
Thanks for the suggestions. I'll definitely remove the generous use of with, but unless I'm missing something, I'll need to keep `pkgs` as it's required for `buildGoModule`.

brhoades replied:
I've figured this out: I can just put required functions in this function's set and it'll work out.

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+  pname = "ytt";
+  version = "0.30.0";
+
+  goPackagePath = "github.com/vmware-tanzu/carvel-ytt";
+  src = fetchFromGitHub {
+    sha256 = "0v9wp15aj4r7wif8i897zwj3c6bg41b95kk7vi3a3bzin814qn6l";
+    rev = "v${version}";
+    repo = "carvel-ytt";
+    owner = "vmware-tanzu";
+  };
+  vendorSha256 = null;
```
```suggestion
  src = fetchFromGitHub {
    owner = "vmware-tanzu";
    repo = "carvel-ytt";
    rev = "v${version}";
    sha256 = "0v9wp15aj4r7wif8i897zwj3c6bg41b95kk7vi3a3bzin814qn6l";
  };

  goPackagePath = "github.com/vmware-tanzu/carvel-ytt";

  vendorSha256 = null;
```

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+  vendorSha256 = null;
+
+  # on v0.30.0 YAML marshalling tests fail without consequence for ytt itself
+  checkPhase = "";
```
```suggestion
  doCheck = false;
```

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+  # on v0.30.0 YAML marshalling tests fail without consequence for ytt itself
+  checkPhase = "";
+  preFixup = ''
+    rm -f $out/bin/ytt-lambda-website
```
```suggestion
    rm $out/bin/ytt-lambda-website
```

Otherwise we won't notice when something changes.

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+  '';
+
+  meta = {
+    description = ''
+      YAML templating tool that allows configuration of complex software via reusable templates with user-provided values
+    '';
+
```
```suggestion
    description = "YAML templating tool that allows configuration of complex software via reusable templates with user-provided values";
```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28655,4 +28655,5 @@ in
 
   psftools = callPackage ../os-specific/linux/psftools {};
 
+  ytt = callPackage ../development/tools/ytt {};
```
Please put this under the tools headline and not at the end of everything.

brhoades replied:
Didn't see the sections on the file, but I now see the file-based sorting. thanks.

brhoades committed:
maintainers: add brodes

brhoades committed:
ytt: init at 0.30.0

Added carvel ytt 0.30.0 which is currently absent. Tests ran by
checkPhase currently fail due to YAML marshalling problems, but the tool
itself appears fine.

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+{ lib, buildGoModule, fetchFromGitHub }:
```
Please update the PR title and commit message to ``ytt: init at 0.30.0``.

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+  meta = with lib; {
+    description = "YAML templating tool that allows configuration of complex software via reusable templates with user-provided values";
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+  vendorSha256 = null;
+
+  # on v0.30.0 YAML marshalling tests fail without consequence for ytt itself
```
It is easily possible to disable this one test?

brhoades replied:
Yep, in fact their [developer documentation](https://github.com/vmware-tanzu/carvel-ytt/blob/develop/docs/dev.md) refers to a script which does just that. Since the script has other dependencies, I've pulled out the important line into a custom checkPhase.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28654,5 +28656,4 @@ in
   cagebreak = callPackage ../applications/window-managers/cagebreak/default.nix {};
 
   psftools = callPackage ../os-specific/linux/psftools {};
-
```
Please revert this change.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 105071` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ytt</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,32 @@
+  '';
+
+  preFixup = ''
+    rm $out/bin/ytt-lambda-website
```
Can we set ``subPackages = [ "cmd/something" ];`` to only build the required binary?

SuperSandro2000 replied:
Even without this change I get:

```
running tests
cannot find package "." in:
        /private/tmp/nix-build-ytt-0.30.0.drv-0/source/cmd/ytt
github.com/k14s/ytt/cmd/ytt-lambda-website
github.com/k14s/ytt/pkg/cmd
github.com/k14s/ytt/pkg/cmd/core
github.com/k14s/ytt/pkg/cmd/template
github.com/k14s/ytt/pkg/filepos
github.com/k14s/ytt/pkg/files
github.com/k14s/ytt/pkg/orderedmap
github.com/k14s/ytt/pkg/structmeta
github.com/k14s/ytt/pkg/template
github.com/k14s/ytt/pkg/template/core
github.com/k14s/ytt/pkg/texttemplate
github.com/k14s/ytt/pkg/version
github.com/k14s/ytt/pkg/website
github.com/k14s/ytt/pkg/workspace
github.com/k14s/ytt/pkg/yamlfmt
github.com/k14s/ytt/pkg/yamlmeta
github.com/k14s/ytt/pkg/yamltemplate
github.com/k14s/ytt/pkg/yttlibrary
github.com/k14s/ytt/pkg/yttlibrary/overlay
```

brhoades replied:
Setting subPackages to `cmd/ytt` has the effect of only building the desired command. We can't run any tests this way, but overall should be faster.

I'm fine with doing it this way and only tried to include tests since they were erroring.

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,32 @@
+    go test "$(go list ./... | grep -v yaml.v2)"
+  '';
+
+  preFixup = ''
+    rm $out/bin/ytt-lambda-website
+  '';
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/tools/ytt/default.nix:
```diff
@@ -0,0 +1,32 @@
+  goPackagePath = "github.com/vmware-tanzu/carvel-ytt";
+
+  vendorSha256 = null;
+
```
```suggestion

subPackages = [ "cmd/ytt" ];

```

brhoades replied:
I'll do this, but there are no tests ran since they're tucked away in other subpackages. It does make for a smaller file overall.

SuperSandro2000 reviewed (commented):
Please change the title and commit message to not contain a v in the version number.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 105071` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ytt</li>
  </ul>
</details>

SuperSandro2000 commented:
Result of `nixpkgs-review pr 105071` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ytt</li>
  </ul>
</details>

brhoades commented:
Change the title and commit.

SuperSandro2000 commented:
Congratulations on your first PR.

brhoades commented:
Thanks for the assistance!
